Failure to build on 32bit Linux: static assertion failed

I am unable to build vvenc on 32bit Linux (Slackware). The Test section and build failure are given below:

```
-- Performing Test SUPPORTED_Werror_unused_command_line_argument
-- Performing Test SUPPORTED_Werror_unused_command_line_argument - Failed
-- Performing Test SUPPORTED_msse4_1
-- Performing Test SUPPORTED_msse4_1 - Success
-- Performing Test SUPPORTED_msse2
-- Performing Test SUPPORTED_msse2 - Success
-- Performing Test SUPPORTED_mavx
-- Performing Test SUPPORTED_mavx - Success
-- Performing Test HAVE_INTRIN_mm_storeu_si16
-- Performing Test HAVE_INTRIN_mm_storeu_si16 - Success
-- Performing Test HAVE_INTRIN_mm_storeu_si32
-- Performing Test HAVE_INTRIN_mm_storeu_si32 - Success
-- Performing Test HAVE_INTRIN_mm_storeu_si64
-- Performing Test HAVE_INTRIN_mm_storeu_si64 - Success
-- Performing Test HAVE_INTRIN_mm_loadu_si32
-- Performing Test HAVE_INTRIN_mm_loadu_si32 - Success
-- Performing Test HAVE_INTRIN_mm_loadu_si64
-- Performing Test HAVE_INTRIN_mm_loadu_si64 - Success
-- Performing Test HAVE_INTRIN_mm256_zeroupper
-- Performing Test HAVE_INTRIN_mm256_zeroupper - Success
-- Performing Test HAVE_INTRIN_mm256_loadu2_m128i
-- Performing Test HAVE_INTRIN_mm256_loadu2_m128i - Success
-- Performing Test HAVE_INTRIN_mm256_set_m128i
-- Performing Test HAVE_INTRIN_mm256_set_m128i - Success
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- ccache found. using it.
-- CMAKE_INSTALL_RPATH=$ORIGIN;$ORIGIN/../lib
-- Performing Test SUPPORTED_mxsave
-- Performing Test SUPPORTED_mxsave - Success
-- Performing Test SUPPORTED_msse4_2
-- Performing Test SUPPORTED_msse4_2 - Success
-- Performing Test SUPPORTED_mavx2
-- Performing Test SUPPORTED_mavx2 - Success
-- Configuring done
-- Generating done
-- Build files have been written to: /tmp/SBo/vvenc-1.7.0/build
[  1%] Building CXX object source/Lib/vvenc/CMakeFiles/vvenc_x86_simd.dir/__/CommonLib/x86/sse41/AdaptiveLoopFilter_sse41.cpp.o
[  1%] Building CXX object source/Lib/vvenc/CMakeFiles/vvenc_x86_simd.dir/__/CommonLib/x86/sse41/AffineGradientSearch_sse41.cpp.o
[  2%] Building CXX object source/Lib/vvenc/CMakeFiles/vvenc_x86_simd.dir/__/CommonLib/x86/sse41/Buffer_sse41.cpp.o
In file included from /tmp/SBo/vvenc-1.7.0/source/Lib/CommonLib/x86/sse41/Buffer_sse41.cpp:43:
/tmp/SBo/vvenc-1.7.0/source/Lib/CommonLib/x86/sse41/../BufferX86.h: In function ‘void vvenc::fillPtrMap_SIMD(void**, ptrdiff_t, int, int, void*)’:
/tmp/SBo/vvenc-1.7.0/source/Lib/CommonLib/x86/sse41/../BufferX86.h:1658:32: error: static assertion failed: Only supported for 64bit systems!
 1658 |   static_assert( sizeof( val ) == 8, "Only supported for 64bit systems!" );
      |                  ~~~~~~~~~~~~~~^~~~
make[2]: *** [source/Lib/vvenc/CMakeFiles/vvenc_x86_simd.dir/build.make:104: source/Lib/vvenc/CMakeFiles/vvenc_x86_simd.dir/__/CommonLib/x86/sse41/Buffer_sse41.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:198: source/Lib/vvenc/CMakeFiles/vvenc_x86_simd.dir/all] Error 2
make: *** [Makefile:146: all] Error 2
root@ilium/home/andrew/Desktop/vvenc$ 
```
Please let me know if I need to add more material to this report.

Andrew

--- a/.github/workflows/Build.yml
+++ b/.github/workflows/Build.yml
@@ -21,13 +21,13 @@ jobs:
               generators: "Visual Studio 16 2019",
               msvc_arch: x64
             }
-#          - {
-#              name: "Windows MSVC 32 Bit",
-#              os: windows-2019,
-#              environment_script: "C:/Program Files (x86)/Microsoft Visual Studio/2019/Enterprise/VC/Auxiliary/Build/vcvars32.bat",
-#              generators: "Visual Studio 16 2019",
-#              msvc_arch: Win32
-#            }
+          - {
+              name: "Windows MSVC 32 Bit",
+              os: windows-2019,
+              environment_script: "C:/Program Files (x86)/Microsoft Visual Studio/2019/Enterprise/VC/Auxiliary/Build/vcvars32.bat",
+              generators: "Visual Studio 16 2019",
+              msvc_arch: Win32
+            }
           - {
               name: "Windows MinGW",
               os: windows-2019,

--- a/.gitlab-ci-internal.yml
+++ b/.gitlab-ci-internal.yml
@@ -47,6 +47,12 @@ variables:
    script:
       - make osx-arch=${OSX_ARCH} disable-lto=1 all
       - make osx-arch=${OSX_ARCH} disable-lto=1 test
+      
+.build_test_msvc_template:
+   extends: .build_test_template
+   script:
+      - make msvc-arch=${MSVC_ARCH} disable-lto=1 all
+      - make msvc-arch=${MSVC_ARCH} disable-lto=1 test
 
 .build_only_ubuntu2004_mingw_template:
    extends: .build_only_template_full
@@ -194,14 +200,32 @@ test_ubuntu2204:
    image: $CI_REGISTRY/pub/dockerimages/ubuntu_2204_full:latest
 
 test_vc191x:
-   extends: .build_test_template
+   extends: .build_test_msvc_template
+   variables:
+      MSVC_ARCH: x64
    tags:
       - vc191x
 
 test_vc192x:
-   extends: .build_test_template
+   extends: .build_test_msvc_template
+   variables:
+      MSVC_ARCH: x64
    tags:
       - vc192x
+      
+test_vc193x_Win32:
+  extends: .build_test_msvc_template
+  variables:
+     MSVC_ARCH: Win32
+  tags:
+    - vc193x
+    
+test_vc193x:
+  extends: .build_test_msvc_template
+  variables:
+     MSVC_ARCH: x64
+  tags:
+    - vc193x
 
 test_macos-x86_64:
    extends: .build_test_macos_template

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -35,7 +35,7 @@ if( VVENC_ENABLE_X86_SIMD )
     add_compile_options( ${FLAG_msse41} )
   endif()
 
-  if( NOT MSVC AND NOT ${CMAKE_SYSTEM_NAME} STREQUAL "Emscripten" )
+  if( NOT ${CMAKE_SYSTEM_NAME} STREQUAL "Emscripten" )
     check_missing_intrinsics()
   endif()
 

--- a/Makefile
+++ b/Makefile
@@ -83,7 +83,11 @@ NUM_JOBS := $(j)
 endif
 
 ifeq ($(OS),Windows_NT)
-  CONFIG_OPTIONS += -A x64
+  ifneq ($(msvc-arch),)
+    CONFIG_OPTIONS += -A $(msvc-arch)
+  else
+    CONFIG_OPTIONS += -A x64
+  endif
   CMAKE_MCONFIG := 1
   ifneq ($(NUM_JOBS),1)
     BUILD_TOOL_OPTIONS := -- /maxcpucount:$(NUM_JOBS) /verbosity:minimal /nr:false

--- a/cmake/modules/vvencCompilerSupport.cmake
+++ b/cmake/modules/vvencCompilerSupport.cmake
@@ -25,8 +25,12 @@ function( check_missing_intrinsics )
   # if building WASM -msimd128 needs to be added to all checks for supported simd-compiler flags
   _emscripten_enable_wasm_simd128()
 
-  set_if_compiler_supports_flag( FLAG_msse2 -msse2 )
-  set_if_compiler_supports_flag( FLAG_mavx  -mavx  )
+  if( NOT MSVC )
+    set_if_compiler_supports_flag( FLAG_msse2 -msse2 )
+    set_if_compiler_supports_flag( FLAG_mavx  -mavx  )
+  else()
+    set_if_compiler_supports_flag( FLAG_mavx  /arch:AVX2  )
+  endif()
 
   # SSE2
   _check_intrinsic( _mm_storeu_si16      "${FLAG_msse2}" "int16_t a = 0; _mm_storeu_si16( &a, _mm_setzero_si128() );"                )
@@ -34,6 +38,8 @@ function( check_missing_intrinsics )
   _check_intrinsic( _mm_storeu_si64      "${FLAG_msse2}" "int64_t a = 0; _mm_storeu_si64( &a, _mm_setzero_si128() );"                )
   _check_intrinsic( _mm_loadu_si32       "${FLAG_msse2}" "int32_t a = 0; __m128i x = _mm_loadu_si32( &a );"                          )
   _check_intrinsic( _mm_loadu_si64       "${FLAG_msse2}" "int64_t a = 0; __m128i x = _mm_loadu_si64( &a );"                          )
+  _check_intrinsic( _mm_cvtsi128_si64    "${FLAG_msse2}" "int64_t a = 0; a = _mm_cvtsi128_si64( _mm_setzero_si128() );"              ) 
+  _check_intrinsic( _mm_extract_epi64    "${FLAG_msse2}" "int64_t a = 0; a = _mm_extract_epi64( _mm_setzero_si128(), 0 );"           )
 
   # AVX
   _check_intrinsic( _mm256_zeroupper     "${FLAG_mavx}"  "_mm256_zeroupper();"                                                       )

--- a/source/Lib/CommonLib/CodingStructure.cpp
+++ b/source/Lib/CommonLib/CodingStructure.cpp
@@ -292,9 +292,8 @@ CodingUnit& CodingStructure::addCU( const UnitArea& unit, const ChannelType chTy
   cus.push_back( cu );
 
   Mv* prevCuMvd = cuInit ? cuInit->mvdL0SubPu : nullptr;
-
-  uint32_t idx = ++m_numCUs;
-  cu->idx  = idx;
+  
+  cu->idx        = ++m_numCUs;
   cu->mvdL0SubPu = nullptr;
 
   if( isLuma( chType ) && unit.lheight() >= 8 && unit.lwidth() >= 8 && unit.Y().area() >= 128 )
@@ -1176,14 +1175,21 @@ const CodingUnit* CodingStructure::getCURestricted( const Position& pos, const C
   const int ydiff  = ( pos.y >> yshift ) - ( curCu.blocks[_chType].y >> yshift );
   const int xdiff  = ( pos.x >> xshift ) - ( curCu.blocks[_chType].x >> xshift );
 
+  if( !xdiff && !ydiff )
+  {
+    const CodingUnit* cu = getCU( pos, _chType, curCu.treeType );
+
+    return ( cu && ( cu->cs != curCu.cs || cu->idx <= curCu.idx ) ) ? cu : nullptr;
+  }
+
   if( ydiff > 0 || ( ydiff == 0 && xdiff > 0 ) || ( ydiff == -1 && xdiff > ( sps->entropyCodingSyncEnabled ? 0 : 1 ) ) )
     return nullptr;
 
-  if( pos.x < 0 || pos.y < 0 || ( pos.x * (1 << csx) ) >= pcv->lumaWidth || pps->getTileIdx( pos.x >> xshift, pos.y >> yshift ) != curCu.tileIdx ) return nullptr;
+  if( pos.x < 0 || pos.y < 0 || ( pos.x * ( 1 << csx ) ) >= pcv->lumaWidth || pps->getTileIdx( pos.x >> xshift, pos.y >> yshift ) != curCu.tileIdx ) return nullptr;
 
-  const CodingUnit* cu = getCU( pos, _chType, curCu.treeType );
+  const CodingUnit* cu = picture->cs->getCU( pos, _chType, curCu.treeType );
 
-  return ( cu && CU::isSameSliceAndTile( *cu, curCu ) && ( cu->cs != curCu.cs || cu->idx <= curCu.idx ) ) ? cu : nullptr;
+  return ( cu && CU::isSameSlice( *cu, curCu ) ) ? cu : nullptr;
 }
 
 const CodingUnit *CodingStructure::getCURestricted( const Position &pos, const Position curPos, const unsigned curSliceIdx, const unsigned curTileIdx, const ChannelType _chType, const TreeType _treeType ) const
@@ -1195,12 +1201,17 @@ const CodingUnit *CodingStructure::getCURestricted( const Position &pos, const P
   const int ydiff  = ( pos.y >> yshift ) - ( curPos.y >> yshift );
   const int xdiff  = ( pos.x >> xshift ) - ( curPos.x >> xshift );
 
+  if( !xdiff && !ydiff )
+  {
+    return getCU( pos, _chType, _treeType );
+  }
+
   if( ydiff > 0 || ( ydiff == 0 && xdiff > 0 ) || ( ydiff == -1 && xdiff > ( sps->entropyCodingSyncEnabled ? 0 : 1 ) ) )
     return nullptr;
 
   if( pos.x < 0 || pos.y < 0 || ( pos.x << csx ) >= pcv->lumaWidth || pps->getTileIdx( pos.x >> xshift, pos.y >> yshift ) != curTileIdx ) return nullptr;
 
-  const CodingUnit* cu = getCU( pos, _chType, _treeType );
+  const CodingUnit* cu = picture->cs->getCU( pos, _chType, _treeType );
 
   return ( cu && cu->slice->independentSliceIdx == curSliceIdx && cu->tileIdx == curTileIdx ) ? cu : nullptr;
 }

--- a/source/Lib/CommonLib/IntraPrediction.cpp
+++ b/source/Lib/CommonLib/IntraPrediction.cpp
@@ -1048,6 +1048,7 @@ int isAboveAvailable(const CodingUnit &cu, const ChannelType& chType, const Posi
       const CodingUnit* cuN = cs.getCURestricted(refPos, cu, chType);
       valid = (cuN != NULL);
       if( cuN ) checkPosX = chType == CH_C ? (cuN->Cb().x + cuN->Cb().width - posLT.x) : (cuN->Y().x + cuN->Y().width - posLT.x);
+      else break;
     }
 
     numIntra += valid ? 1 : 0;
@@ -1078,7 +1079,9 @@ int isLeftAvailable(const CodingUnit &cu, const ChannelType& chType, const Posit
       const CodingUnit* cuN = cs.getCURestricted(refPos, cu, chType);
       valid = (cuN != NULL);
       if( cuN ) checkPosY = chType == CH_C ? (cuN->Cb().y + cuN->Cb().height - posLT.y) : (cuN->Y().y + cuN->Y().height - posLT.y);
+      else break;
     }
+
     numIntra += valid ? 1 : 0;
     *validFlags = valid;
 
@@ -1107,7 +1110,9 @@ int isAboveRightAvailable(const CodingUnit &cu, const ChannelType& chType, const
       const CodingUnit* cuN = cs.getCURestricted(refPos, cu, chType);
       valid = (cuN != NULL);
       if(cuN) checkPosX = chType == CH_C ? (cuN->Cb().x + cuN->Cb().width - (posRT.x + unitWidth)) : (cuN->Y().x + cuN->Y().width - (posRT.x + unitWidth));
+      else break;
     }
+
     numIntra += valid ? 1 : 0;
     *validFlags = valid;
 
@@ -1136,7 +1141,9 @@ int isBelowLeftAvailable(const CodingUnit &cu, const ChannelType& chType, const 
       const CodingUnit* cuN = cs.getCURestricted(refPos, cu, chType);
       valid = (cuN != NULL);
       if( cuN ) checkPosY = chType == CH_C ? (cuN->Cb().y + cuN->Cb().height - (posLB.y + unitHeight)) : (cuN->Y().y + cuN->Y().height - (posLB.y + unitHeight));
+      else break;
     }
+
     numIntra += valid ? 1 : 0;
     *validFlags = valid;
 

--- a/source/Lib/CommonLib/x86/BufferX86.h
+++ b/source/Lib/CommonLib/x86/BufferX86.h
@@ -1652,10 +1652,11 @@ void applyLut_SIMD( const Pel* src, const ptrdiff_t srcStride, Pel* dst, const p
   }
 }
 
+#if INTPTR_MAX == INT64_MAX
 template<X86_VEXT vext>
 void fillPtrMap_SIMD( void** ptr, ptrdiff_t ptrStride, int width, int height, void* val )
 {
-  static_assert( sizeof( val ) == 8, "Only supported for 64bit systems!" );
+  static_assert( sizeof( ptr ) == 8, "Only supported for 64bit systems!" );
   if( ( width & 3 ) == 0 )
   {
 #if USE_AVX2
@@ -1701,6 +1702,77 @@ void fillPtrMap_SIMD( void** ptr, ptrdiff_t ptrStride, int width, int height, vo
     }
   }
 }
+#elif INTPTR_MAX == INT32_MAX
+template<X86_VEXT vext>
+void fillPtrMap_SIMD( void** ptr, ptrdiff_t ptrStride, int width, int height, void* val )
+{
+  static_assert( sizeof( ptr ) == 4, "Only supported for 32bit systems!" );
+  if( ( width & 7 ) == 0 )
+  {
+#if USE_AVX2
+    __m256i vval = _mm256_set1_epi32( ( int32_t ) val );
+
+    while( height-- )
+    {
+      for( int x = 0; x < width; x += 8 )
+      {
+        _mm256_storeu_si256( ( __m256i* ) &ptr[x], vval );
+      }
+
+      ptr += ptrStride;
+    }
+#else
+    __m128i vval = _mm_set1_epi32( ( int32_t ) val );
+
+    while( height-- )
+    {
+      for( int x = 0; x < width; x += 8 )
+      {
+        _mm_storeu_si128( ( __m128i* ) &ptr[x + 0], vval );
+        _mm_storeu_si128( ( __m128i* ) &ptr[x + 4], vval );
+      }
+
+      ptr += ptrStride;
+    }
+#endif
+  }
+  else if( ( width & 3 ) == 0 )
+  {
+    __m128i vval = _mm_set1_epi32( ( int32_t ) val );
+
+    while( height-- )
+    {
+      for( int x = 0; x < width; x += 4 )
+      {
+        _mm_storeu_si128( ( __m128i* ) &ptr[x], vval );
+      }
+
+      ptr += ptrStride;
+    }
+  }
+  else if( ( width & 1 ) == 0 )
+  {
+    while( height-- )
+    {
+      ptr[0] = val;
+      ptr[1] = val;
+
+      ptr += ptrStride;
+    }
+  }
+  else
+  {
+    while( height-- )
+    {
+      for( int x = 0; x < width; ++x )
+      {
+        ptr[x] = val;
+      }
+      ptr += ptrStride;
+    }
+  }
+}
+#endif  // INTPTR_MAX == INT32_MAX
 
 template<X86_VEXT vext>
 uint64_t AvgHighPass_SIMD( const int width, const int height, const Pel* pSrc, const int iSrcStride)

--- a/source/Lib/CommonLib/x86/FixMissingIntrin.h
+++ b/source/Lib/CommonLib/x86/FixMissingIntrin.h
@@ -92,6 +92,32 @@ static inline __m128i _mm_loadu_si64( const void* p )
 }
 #endif
 
+#ifdef MISSING_INTRIN_mm_cvtsi128_si64
+#if INTPTR_MAX == INT64_MAX
+#error __mm_cvtsi128_si64 has to be defined for 64-bit systems!
+#endif
+static inline int64_t _mm_cvtsi128_si64( __m128i a )
+{
+  int64_t x;
+  _mm_storel_epi64( ( __m128i* ) &x, a );
+  return x;
+}
+#endif
+
+#ifdef MISSING_INTRIN_mm_extract_epi64
+#if INTPTR_MAX == INT64_MAX
+#error _mm_extract_epi64 has to be defined for 64-bit systems!
+#endif
+static inline int64_t _mm_extract_epi64( __m128i a, int i )
+{
+  int64_t x;
+  if( i )
+    _mm_storel_epi64( ( __m128i* ) &x, _mm_unpackhi_epi64( a, a ) );
+  else
+    _mm_storel_epi64( ( __m128i* ) &x, a );
+  return x;
+}
+#endif
 
 #if defined( USE_AVX ) || defined( USE_AVX2 )
 

--- a/source/Lib/CommonLib/x86/InitX86.cpp
+++ b/source/Lib/CommonLib/x86/InitX86.cpp
@@ -94,8 +94,10 @@ void InterpolationFilter::initInterpolationFilterX86( /*int iBitDepthY, int iBit
 #if ENABLE_SIMD_OPT_BUFFER
 void PelBufferOps::initPelBufOpsX86()
 {
-  if ( isInitX86Done )
+  if( isInitX86Done )
     return;
+
+  isInitX86Done = true;
 
   auto vext = read_x86_extension_flags();
   switch (vext){
@@ -113,8 +115,6 @@ void PelBufferOps::initPelBufOpsX86()
     default:
       break;
   }
-
-  isInitX86Done = true;
 }
 #endif
 
@@ -207,7 +207,6 @@ void SampleAdaptiveOffset::initSampleAdaptiveOffsetX86()
     default:
       break;
   }
-
 }
 
 #endif

--- a/source/Lib/CommonLib/x86/InterpolationFilterX86.h
+++ b/source/Lib/CommonLib/x86/InterpolationFilterX86.h
@@ -1155,141 +1155,8 @@ static void simdInterpolateVerM16_AVX2( const int16_t *src, int srcStride, int16
 #endif
 }
 
-
-template<int N, bool isLast>
-inline void interpolate( const int16_t* src, int cStride, int16_t *dst, int width, int shift, int offset, int bitdepth, int maxVal, int16_t const *c )
-{
-  for( int col = 0; col < width; col++ )
-  {
-    int sum;
-
-    sum = src[col + 0 * cStride] * c[0];
-    sum += src[col + 1 * cStride] * c[1];
-    if( N >= 4 )
-    {
-      sum += src[col + 2 * cStride] * c[2];
-      sum += src[col + 3 * cStride] * c[3];
-    }
-    if( N >= 6 )
-    {
-      sum += src[col + 4 * cStride] * c[4];
-      sum += src[col + 5 * cStride] * c[5];
-    }
-    if( N == 8 )
-    {
-      sum += src[col + 6 * cStride] * c[6];
-      sum += src[col + 7 * cStride] * c[7];
-    }
-
-    Pel val = ( sum + offset ) >> shift;
-    if( isLast )
-    {
-      val = ( val < 0 ) ? 0 : val;
-      val = ( val > maxVal ) ? maxVal : val;
-    }
-    dst[col] = val;
-  }
-}
-
-
-static inline __m128i simdInterpolateLuma2P8( int16_t const *src, int srcStride, __m128i *mmCoeff, const __m128i & mmOffset, int shift )
-{
-  __m128i sumHi = _mm_setzero_si128();
-  __m128i sumLo = _mm_setzero_si128();
-  for( int n = 0; n < 2; n++ )
-  {
-    __m128i mmPix = _mm_loadu_si128( ( __m128i* )src );
-    __m128i hi = _mm_mulhi_epi16( mmPix, mmCoeff[n] );
-    __m128i lo = _mm_mullo_epi16( mmPix, mmCoeff[n] );
-    sumHi = _mm_add_epi32( sumHi, _mm_unpackhi_epi16( lo, hi ) );
-    sumLo = _mm_add_epi32( sumLo, _mm_unpacklo_epi16( lo, hi ) );
-    src += srcStride;
-  }
-  sumHi = _mm_srai_epi32( _mm_add_epi32( sumHi, mmOffset ), shift );
-  sumLo = _mm_srai_epi32( _mm_add_epi32( sumLo, mmOffset ), shift );
-  return( _mm_packs_epi32( sumLo, sumHi ) );
-}
-
-static inline __m128i simdInterpolateLuma2P4( int16_t const *src, int srcStride, __m128i *mmCoeff, const __m128i & mmOffset, int shift )
-{
-  __m128i sumHi = _mm_setzero_si128();
-  __m128i sumLo = _mm_setzero_si128();
-  for( int n = 0; n < 2; n++ )
-  {
-    __m128i mmPix = _mm_loadl_epi64( ( __m128i* )src );
-    __m128i hi = _mm_mulhi_epi16( mmPix, mmCoeff[n] );
-    __m128i lo = _mm_mullo_epi16( mmPix, mmCoeff[n] );
-    sumHi = _mm_add_epi32( sumHi, _mm_unpackhi_epi16( lo, hi ) );
-    sumLo = _mm_add_epi32( sumLo, _mm_unpacklo_epi16( lo, hi ) );
-    src += srcStride;
-  }
-  sumHi = _mm_srai_epi32( _mm_add_epi32( sumHi, mmOffset ), shift );
-  sumLo = _mm_srai_epi32( _mm_add_epi32( sumLo, mmOffset ), shift );
-  return( _mm_packs_epi32( sumLo, sumHi ) );
-}
-
-static inline __m128i simdClip3( __m128i mmMin, __m128i mmMax, __m128i mmPix )
-{
-  __m128i mmMask = _mm_cmpgt_epi16( mmPix, mmMin );
-  mmPix = _mm_or_si128( _mm_and_si128( mmMask, mmPix ), _mm_andnot_si128( mmMask, mmMin ) );
-  mmMask = _mm_cmplt_epi16( mmPix, mmMax );
-  mmPix = _mm_or_si128( _mm_and_si128( mmMask, mmPix ), _mm_andnot_si128( mmMask, mmMax ) );
-  return( mmPix );
-}
-
-template<X86_VEXT vext, bool isLast>
-static void simdInterpolateN2_M8( const int16_t* src, int srcStride, int16_t *dst, int dstStride, int cStride, int width, int height, int shift, int offset, const ClpRng& clpRng, int16_t const *c )
-{
-  int row, col;
-  __m128i mmOffset = _mm_set1_epi32( offset );
-  __m128i mmCoeff[2];
-  __m128i mmMin = _mm_set1_epi16( clpRng.min() );
-  __m128i mmMax = _mm_set1_epi16( clpRng.max() );
-  for( int n = 0; n < 2; n++ )
-    mmCoeff[n] = _mm_set1_epi16( c[n] );
-  for( row = 0; row < height; row++ )
-  {
-    for( col = 0; col < width; col += 8 )
-    {
-      __m128i mmFiltered = simdInterpolateLuma2P8( src + col, cStride, mmCoeff, mmOffset, shift );
-      if( isLast )
-      {
-        mmFiltered = simdClip3( mmMin, mmMax, mmFiltered );
-      }
-      _mm_storeu_si128( ( __m128i * )( dst + col ), mmFiltered );
-    }
-    src += srcStride;
-    dst += dstStride;
-  }
-}
-
-template<X86_VEXT vext, bool isLast>
-static void simdInterpolateN2_M4( const int16_t* src, int srcStride, int16_t *dst, int dstStride, int cStride, int width, int height, int shift, int offset, const ClpRng& clpRng, int16_t const *c )
-{
-  int row, col;
-  __m128i mmOffset = _mm_set1_epi32( offset );
-  __m128i mmCoeff[8];
-  __m128i mmMin = _mm_set1_epi16( clpRng.min() );
-  __m128i mmMax = _mm_set1_epi16( clpRng.max() );
-  for( int n = 0; n < 2; n++ )
-    mmCoeff[n] = _mm_set1_epi16( c[n] );
-  for( row = 0; row < height; row++ )
-  {
-    for( col = 0; col < width; col += 4 )
-    {
-      __m128i mmFiltered = simdInterpolateLuma2P4( src + col, cStride, mmCoeff, mmOffset, shift );
-      if( isLast )
-      {
-        mmFiltered = simdClip3( mmMin, mmMax, mmFiltered );
-      }
-      _mm_storel_epi64( ( __m128i * )( dst + col ), mmFiltered );
-    }
-    src += srcStride;
-    dst += dstStride;
-  }
-}
 #ifdef USE_AVX2
-static inline __m256i simdInterpolateLuma10Bit2P16(int16_t const *src1, int srcStride, __m256i *mmCoeff, const __m256i & mmOffset, __m128i &mmShift)
+static inline __m256i simdInterpolateLuma10Bit2P16(int16_t const *src1, int srcStride, __m256i *mmCoeff, const __m256i & mmOffset, int shift)
 {
   __m256i sumLo;
   {
@@ -1299,12 +1166,12 @@ static inline __m256i simdInterpolateLuma10Bit2P16(int16_t const *src1, int srcS
     __m256i lo1 = _mm256_mullo_epi16(mmPix1, mmCoeff[1]);
     sumLo = _mm256_add_epi16(lo0, lo1);
   }
-  sumLo = _mm256_sra_epi16(_mm256_add_epi16(sumLo, mmOffset), mmShift);
+  sumLo = _mm256_srai_epi16(_mm256_add_epi16(sumLo, mmOffset), shift);
   return(sumLo);
 }
 #endif
 
-static inline __m128i simdInterpolateLuma10Bit2P8(int16_t const *src1, int srcStride, __m128i *mmCoeff, const __m128i & mmOffset, __m128i &mmShift)
+static inline __m128i simdInterpolateLuma10Bit2P8(int16_t const *src1, int srcStride, __m128i *mmCoeff, const __m128i & mmOffset, int shift)
 {
   __m128i sumLo;
   {
@@ -1314,11 +1181,11 @@ static inline __m128i simdInterpolateLuma10Bit2P8(int16_t const *src1, int srcSt
     __m128i lo1 = _mm_mullo_epi16(mmPix1, mmCoeff[1]);
     sumLo = _mm_add_epi16(lo0, lo1);
   }
-  sumLo = _mm_sra_epi16(_mm_add_epi16(sumLo, mmOffset), mmShift);
+  sumLo = _mm_srai_epi16(_mm_add_epi16(sumLo, mmOffset), shift);
   return(sumLo);
 }
 
-static inline __m128i simdInterpolateLuma10Bit2P4(int16_t const *src, int srcStride, __m128i *mmCoeff, const __m128i & mmOffset, __m128i &mmShift)
+static inline __m128i simdInterpolateLuma10Bit2P4(int16_t const *src, int srcStride, __m128i *mmCoeff, const __m128i & mmOffset, int shift)
 {
   __m128i sumLo;
   {
@@ -1328,7 +1195,7 @@ static inline __m128i simdInterpolateLuma10Bit2P4(int16_t const *src, int srcStr
     __m128i lo1 = _mm_mullo_epi16(mmPix1, mmCoeff[1]);
     sumLo = _mm_add_epi16(lo0, lo1);
   }
-  sumLo = _mm_sra_epi16(_mm_add_epi16(sumLo, mmOffset), mmShift);
+  sumLo = _mm_srai_epi16(_mm_add_epi16(sumLo, mmOffset), shift);
   return sumLo;
 }
 
@@ -1337,7 +1204,6 @@ static void simdInterpolateN2_10BIT_M4(const int16_t* src, int srcStride, int16_
 {
   int row, col;
   __m128i mmOffset = _mm_set1_epi16(offset);
-  __m128i mmShift = _mm_set_epi64x(0, shift);
   __m128i mmCoeff[2];
   for (int n = 0; n < 2; n++)
     mmCoeff[n] = _mm_set1_epi16(c[n]);
@@ -1357,19 +1223,19 @@ static void simdInterpolateN2_10BIT_M4(const int16_t* src, int srcStride, int16_
     // multiple of 16
     for (; col < ((width >> 4) << 4); col += 16)
     {
-      __m256i mmFiltered = simdInterpolateLuma10Bit2P16(src + col, cStride, mm256Coeff, mm256Offset, mmShift);
+      __m256i mmFiltered = simdInterpolateLuma10Bit2P16(src + col, cStride, mm256Coeff, mm256Offset, shift);
       _mm256_storeu_si256((__m256i *)(dst + col), mmFiltered);
     }
 #endif
     // multiple of 8
     for (; col < ((width >> 3) << 3); col += 8)
     {
-      __m128i mmFiltered = simdInterpolateLuma10Bit2P8(src + col, cStride, mmCoeff, mmOffset, mmShift);
+      __m128i mmFiltered = simdInterpolateLuma10Bit2P8(src + col, cStride, mmCoeff, mmOffset, shift);
       _mm_storeu_si128((__m128i *)(dst + col), mmFiltered);
     }
 
     // last 4 samples
-    __m128i mmFiltered = simdInterpolateLuma10Bit2P4(src + col, cStride, mmCoeff, mmOffset, mmShift);
+    __m128i mmFiltered = simdInterpolateLuma10Bit2P4(src + col, cStride, mmCoeff, mmOffset, shift);
     _mm_storel_epi64((__m128i *)(dst + col), mmFiltered);
     src += srcStride;
     dst += dstStride;
@@ -1502,8 +1368,7 @@ static void simdFilter( const ClpRng& clpRng, Pel const *src, int srcStride, Pel
 
       THROW( "Unhandled case!" );
     }
-
-    if( N == 8 && !( width & 0x07 ) )
+    else if( N == 8 && !( width & 0x07 ) )
     {
       if( !isVertical )
       {
@@ -1579,23 +1444,17 @@ static void simdFilter( const ClpRng& clpRng, Pel const *src, int srcStride, Pel
       }
       return;
     }
-    else if (biMCForDMVR)
+    else if( biMCForDMVR )
     {
-      if (N == 2 && !(width & 0x03))
+      if( N == 2 && !( width & 0x03 ) )
       {
-        simdInterpolateN2_10BIT_M4<vext, isLast>(src, srcStride, dst, dstStride, cStride, width, height, shift, offset, clpRng, c);
+        simdInterpolateN2_10BIT_M4<vext, isLast>( src, srcStride, dst, dstStride, cStride, width, height, shift, offset, clpRng, c );
         return;
       }
     }
-    else if( N == 2 && !( width & 0x07 ) )
+    else if( N == 2 )
     {
-      simdInterpolateN2_M8<vext, isLast>( src, srcStride, dst, dstStride, cStride, width, height, shift, offset, clpRng, c );
-      return;
-    }
-    else if( N == 2 && !( width & 0x03 ) )
-    {
-      simdInterpolateN2_M4<vext, isLast>( src, srcStride, dst, dstStride, cStride, width, height, shift, offset, clpRng, c );
-      return;
+      THROW( "Should have already been handled!" );
     }
     else if( N == 8 && width == 1 && ( height & 3 ) == 0 && !isVertical )
     {
@@ -3233,61 +3092,61 @@ void InterpolationFilter::_initInterpolationFilterX86()
   m_filterHor[0][0][1] = simdFilter<vext, 8, false, false, true>;
   m_filterHor[0][1][0] = simdFilter<vext, 8, false, true, false>;
   m_filterHor[0][1][1] = simdFilter<vext, 8, false, true, true>;
-
+  
   m_filterHor[1][0][0] = simdFilter<vext, 4, false, false, false>;
   m_filterHor[1][0][1] = simdFilter<vext, 4, false, false, true>;
   m_filterHor[1][1][0] = simdFilter<vext, 4, false, true, false>;
   m_filterHor[1][1][1] = simdFilter<vext, 4, false, true, true>;
-
+  
   m_filterHor[2][0][0] = simdFilter<vext, 2, false, false, false>;
   m_filterHor[2][0][1] = simdFilter<vext, 2, false, false, true>;
   m_filterHor[2][1][0] = simdFilter<vext, 2, false, true, false>;
   m_filterHor[2][1][1] = simdFilter<vext, 2, false, true, true>;
-
+  
   m_filterHor[3][0][0] = simdFilter<vext, 6, false, false, false>;
   m_filterHor[3][0][1] = simdFilter<vext, 6, false, false, true>;
   m_filterHor[3][1][0] = simdFilter<vext, 6, false, true, false>;
   m_filterHor[3][1][1] = simdFilter<vext, 6, false, true, true>;
-
+  
   m_filterVer[0][0][0] = simdFilter<vext, 8, true, false, false>;
   m_filterVer[0][0][1] = simdFilter<vext, 8, true, false, true>;
   m_filterVer[0][1][0] = simdFilter<vext, 8, true, true, false>;
   m_filterVer[0][1][1] = simdFilter<vext, 8, true, true, true>;
-
+  
   m_filterVer[1][0][0] = simdFilter<vext, 4, true, false, false>;
   m_filterVer[1][0][1] = simdFilter<vext, 4, true, false, true>;
   m_filterVer[1][1][0] = simdFilter<vext, 4, true, true, false>;
   m_filterVer[1][1][1] = simdFilter<vext, 4, true, true, true>;
-
+  
   m_filterVer[2][0][0] = simdFilter<vext, 2, true, false, false>;
   m_filterVer[2][0][1] = simdFilter<vext, 2, true, false, true>;
   m_filterVer[2][1][0] = simdFilter<vext, 2, true, true, false>;
   m_filterVer[2][1][1] = simdFilter<vext, 2, true, true, true>;
-
+  
   m_filterVer[3][0][0] = simdFilter<vext, 6, true, false, false>;
   m_filterVer[3][0][1] = simdFilter<vext, 6, true, false, true>;
   m_filterVer[3][1][0] = simdFilter<vext, 6, true, true, false>;
   m_filterVer[3][1][1] = simdFilter<vext, 6, true, true, true>;
-
+  
   m_filterCopy[0][0]   = simdFilterCopy<vext, false, false>;
   m_filterCopy[0][1]   = simdFilterCopy<vext, false, true>;
   m_filterCopy[1][0]   = simdFilterCopy<vext, true, false>;
   m_filterCopy[1][1]   = simdFilterCopy<vext, true, true>;
-
+  
   m_filter4x4[0][0]    = simdFilter4x4_N6<vext, false>;
   m_filter4x4[0][1]    = simdFilter4x4_N6<vext, true>;
   m_filter4x4[1][0]    = simdFilter4x4_N4<vext, false>;
   m_filter4x4[1][1]    = simdFilter4x4_N4<vext, true>;
-
+  
   m_filter8x8[0][0]    = simdFilter8xX_N8<vext, false>;
   m_filter8x8[0][1]    = simdFilter8xX_N8<vext, true>;
   m_filter8x8[1][0]    = simdFilter8xX_N4<vext, false>;
   m_filter8x8[1][1]    = simdFilter8xX_N4<vext, true>;
-
-
+  
+  
   m_filter16x16[0][0]    = simdFilter16xX_N8<vext, false>;
   m_filter16x16[0][1]    = simdFilter16xX_N8<vext, true>;
-
+  
   m_filter16x16[1][0]    = simdFilter16xX_N4<vext, false>;
   m_filter16x16[1][1]    = simdFilter16xX_N4<vext, true>;
 

--- a/source/Lib/CommonLib/x86/LoopFilterX86.h
+++ b/source/Lib/CommonLib/x86/LoopFilterX86.h
@@ -669,7 +669,9 @@ template <X86_VEXT vext>
 void LoopFilter::_initLoopFilterX86()
 {
   xPelFilterLuma  = xPelFilterLumaX86<vext>;
+#if INTPTR_MAX == INT64_MAX || !defined( _WIN32 )
   xFilteringPandQ = xFilteringPandQX86<vext>;
+#endif
 }
 
 template void LoopFilter::_initLoopFilterX86<SIMDX86>();

--- a/source/Lib/CommonLib/x86/RdCostX86.h
+++ b/source/Lib/CommonLib/x86/RdCostX86.h
@@ -93,7 +93,7 @@ Distortion RdCost::xGetSSE_SIMD( const DistParam &rcDtParam )
     }
     Sum = _mm256_hadd_epi32( Sum, Sum );
     Sum = _mm256_hadd_epi32( Sum, Sum );
-    uiRet = ( _mm_cvtsi128_si64( _mm256_castsi256_si128( Sum ) ) + _mm_cvtsi128_si32( _mm256_castsi256_si128( _mm256_permute2x128_si256( Sum, Sum, 0x11 ) ) ) ) >> uiShift;
+    uiRet = ( _mm_cvtsi128_si32( _mm256_castsi256_si128( Sum ) ) + _mm_cvtsi128_si32( _mm256_castsi256_si128( _mm256_permute2x128_si256( Sum, Sum, 0x11 ) ) ) ) >> uiShift;
 #endif
   }
   else if( ( iCols & 7 ) == 0 )
@@ -114,7 +114,7 @@ Distortion RdCost::xGetSSE_SIMD( const DistParam &rcDtParam )
     }
     Sum = _mm_hadd_epi32( Sum, Sum );
     Sum = _mm_hadd_epi32( Sum, Sum );
-    uiRet = _mm_cvtsi128_si64( Sum )>>uiShift;
+    uiRet = _mm_cvtsi128_si32( Sum )>>uiShift;
   }
   else
   {
@@ -133,7 +133,7 @@ Distortion RdCost::xGetSSE_SIMD( const DistParam &rcDtParam )
       pSrc2   += iStrideSrc2;
     }
     Sum = _mm_hadd_epi32( Sum, Sum );
-    uiRet = _mm_cvtsi128_si64( Sum )>>uiShift;
+    uiRet = _mm_cvtsi128_si32( Sum )>>uiShift;
   }
 
   return uiRet;
@@ -166,7 +166,7 @@ Distortion RdCost::xGetSSE_NxN_SIMD( const DistParam &rcDtParam )
       Sum = _mm_add_epi32( Sum, Res );
     }
     Sum = _mm_hadd_epi32( Sum, Sum );
-    uiRet = _mm_cvtsi128_si64( Sum )>>uiShift;
+    uiRet = _mm_cvtsi128_si32( Sum )>>uiShift;
   }
   else
   {

--- a/source/Lib/EncoderLib/RateCtrl.cpp
+++ b/source/Lib/EncoderLib/RateCtrl.cpp
@@ -1044,19 +1044,6 @@ void RateCtrl::initRateControlPic( Picture& pic, Slice* slice, int& qp, double& 
           d = firstPassSliceQP - ( 105.0 / 128.0 ) * sqrt( (double)std::max( 1, firstPassSliceQP ) ) * log( d ) / log( 2.0 );
           sliceQP = int( 0.5 + d + 0.5 * std::max( 0.0, baseQP - d ) + encRCSeq->qpCorrection[ frameLevel ] );
 
-#if 0 // TODO hlm: check if this is still needed
-          if ( it->refreshParameters ) // avoid overcoding after some scene cuts (stabilization)
-          {
-            const int offset = ( it->poc > 0 && ( m_pcEncCfg->m_PadSourceWidth > 2048 || m_pcEncCfg->m_PadSourceHeight > 1280 ) ? 5 : 4 ) << ( encRCSeq->bitDepth - 2 ); // to compensate for downsampling with UHD
-            const int clipQP = ( ( ( offset + ( it->poc > 0 ? visAct : 1 << ( encRCSeq->bitDepth - 1 ) ) ) * secondPassBaseQP ) >> ( encRCSeq->bitDepth + 1 ) ) + ( it->isIntra ? m_pcEncCfg->m_intraQPOffset : 0 );
-
-            if ( sliceQP < clipQP )
-            {
-              encRcPic->targetBits = int( 0.5 + encRcPic->targetBits * pow( 2.0, (sliceQP - clipQP) / 5.0 ) ); // VCIP paper, Tab. 1
-              sliceQP = clipQP;
-            }
-          }
-#endif
           encRcPic->clipTargetQP( getPicList(), ( m_pcEncCfg->m_LookAhead ? getBaseQP() : secondPassBaseQP + ( it->isIntra ? m_pcEncCfg->m_intraQPOffset : 0 ) ), sliceQP );
           lambda = it->lambda * pow( 2.0, double( sliceQP - firstPassSliceQP ) / 3.0 );
           lambda = Clip3( encRcSeq->minEstLambda, encRcSeq->maxEstLambda, lambda );

--- a/test/vvenclibtest/CMakeLists.txt
+++ b/test/vvenclibtest/CMakeLists.txt
@@ -31,4 +31,4 @@ source_group( "Resource Files" FILES ${RESOURCE_FILE} )
 
 
 # set the folder where to place the projects
-set_target_properties( ${EXE_NAME}  PROPERTIES FOLDER app )
+set_target_properties( ${EXE_NAME}  PROPERTIES FOLDER test )

--- a/test/vvenclibtest/vvenclibtest.cpp
+++ b/test/vvenclibtest/vvenclibtest.cpp
@@ -1004,7 +1004,6 @@ int checkTimestampsDefault()
 
       if( 0 != runEncoder(c,frames) )
       {
-        std::cout << "Wrong a" << std::endl;
         return -1;
       }
     }

--- a/test/vvenclibtest/vvenclibtest.cpp
+++ b/test/vvenclibtest/vvenclibtest.cpp
@@ -966,11 +966,13 @@ static int runEncoder( vvenc_config& c, uint64_t framesToEncode )
 
   vvenc_YUVBuffer_free( yuvPicture, true );
   vvenc_accessUnit_free( AU, true );
+  vvenc_encoder_close( enc );
   return 0;
 
 fail:
   vvenc_YUVBuffer_free( yuvPicture, true );
   vvenc_accessUnit_free( AU, true );
+  vvenc_encoder_close( enc );
   return -1;
 }
 
@@ -1002,6 +1004,7 @@ int checkTimestampsDefault()
 
       if( 0 != runEncoder(c,frames) )
       {
+        std::cout << "Wrong a" << std::endl;
         return -1;
       }
     }

